runatlantis.io/blog/terraform-and-the-dangers-of-applying-locally.md

```diff
@@ -0,0 +1,120 @@
+---
+title: Terraform And The Dangers Of Applying Locally
+lang: en-US
+---
+
+# Terraform And The Dangers Of Applying Locally
+
+::: info
+This post was originally written on July 13th, 2018
+
+Original post: <https://medium.com/runatlantis/terraform-and-the-dangers-of-applying-locally-543563782a73>
+:::
+
+If you're using Terraform then at some point you've likely ran a `terraform apply` that reverted someone else's change!
+
+Here's how that tends to happen:
+
+## The Setup
+
+Say we have two developers: Alice and Bob. Alice needs to add a new security group rule. She checks out a new branch, adds her rule and creates a pull request:
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic1.webp)
+
+When she runs `terraform plan` locally she sees what she expects.
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic2.webp)
+
+Meanwhile, Bob is working on an emergency fix. He checks out a new branch and adds a different security group rule called `emergency`:
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic3.webp)
+
+And, because it's an emergency, he **immediately runs apply**:
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic4.webp)
+
+Now back to Alice. She's just gotten approval on her pull request change and so she runs `terraform apply`:
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic5.webp)
+
+Did you catch what happened? Did you notice that the `apply` deleted Bob's rule?
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic6.webp)
+
+In this example, it wasn't too hard to see. However if the plan is much longer, or if the change is less obvious then it can be easy to miss.
+
+## Possible Solutions
+
+There are some ways to avoid this:
+
+### Use terraform plan `-out`
+
+If Alice had run `terraform plan -out plan.tfplan` then when she ran `terraform apply plan.tfplan` she would see:
+
+![](/blog/terraform-and-the-dangers-of-applying-locally/pic7.webp)
+
+The problem with this solution is that few people run `terraform plan` anymore, much less `terraform plan -out`!
+
+<iframe src="https://cdn.embedly.com/widgets/media.html?type=text%2Fhtml&amp;key=a19fcc184b9711e1b4764040d3dc5c07&amp;schema=twitter&amp;url=https%3A//twitter.com/sethvargo/status/989979940098424832&amp;image=https%3A//i.embed.ly/1/image%3Furl%3Dhttps%253A%252F%252Fpbs.twimg.com%252Fprofile_images%252F808025120296013825%252FfrGuc14s_400x400.jpg%26key%3Da19fcc184b9711e1b4764040d3dc5c07" allowfullscreen="" frameborder="0" height="249" width="680" title="Seth Vargo on Twitter" class="fr n gh dv bg" scrolling="no"></iframe>
```

Comment:
this iframe embed is broken (https://deploy-preview-4569--runatlantis.netlify.app/blog/terraform-and-the-dangers-of-applying-locally.html#use-terraform-plan-out)

Reply:
that is true, I saw that as well, I am gonna update csp (but I also think of removing it as well)

Reply:
I'm +1 on removing CSP; I don't think we're a likely target for bad actors since we have no PII, credentials of any kind, and is entirely stateless on the website. 

The primary thing I could see being useful for targeting the site is to hijack the "copy code" button in code-fences and manipulate/inject different code than what shown on the site, hoping folks just blindly pastes it into a system. 

I'm not a security professional, so can't say how credible that _actually_ is. 

Alternatively, we can also actively decide NOT to embed anything, but instead provide a static screenshots and link to the source; keeping CSP in place, not have it be ever expanding in scope, and embrace the full `belt and braces` mindset in locking down the site.

Reply:
Embedding has a risk of breaking, degrading the content experience and "history" of the site over the years - which is why I personally take the screenshot + external link approach to ensure my site is durable against churn of other sites over time :)

Reply:
yeah, for now, I am gonna remove the csp and see it goes.